addons/purchase/migrations/9.0.1.2/post-migration.py

```diff
@@ -92,9 +94,182 @@ def account_properties(cr):
+                    JOIN account_invoice ai ON ai.id = ail.invoice_id
+                    JOIN product_uom pol2_u ON pol2_u.id = ail.uom_id
+                    JOIN product_uom ail_u ON ail_u.id = pol2.product_uom
+                WHERE
+                    ai.state != 'cancel' AND
+                    ai.type IN ('in_invoice', 'in_refund')
+                GROUP BY pol2.id
+            ) AS sub
+            WHERE pol.id = sub.id
+        """
+    )
+    # Replace https://bit.ly/369kUDt and leave the rest of the method to
+    # be called in end-migration, where mrp module will be available for
+    # sure in the environment if it is installed
+    openupgrade.logged_query(
+        env.cr,
+        """
+            UPDATE purchase_order_line pol
+            SET qty_received = 0.0
+            FROM purchase_order po
+            WHERE
+                po.id = pol.order_id AND
+                po.state NOT IN ('purchase', 'done')
+        """
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+            UPDATE purchase_order_line pol
+            SET qty_received = pol.product_qty
+            FROM purchase_order_line pol2
+            JOIN purchase_order po ON pol2.order_id = po.id
+            JOIN product_product pp ON pol2.product_id = pp.id
+            JOIN product_template pt ON pp.product_tmpl_id = pt.id
+            WHERE
+                po.id = pol2.id AND
+                po.state IN ('purchase', 'done') AND
+                pt.type NOT IN ('consu', 'product')
+        """
+    )
+    # Replace https://bit.ly/3832SEF
+    openupgrade.logged_query(
+        env.cr,
+        """
+            UPDATE purchase_order po
+            SET invoice_status = CASE
+                WHEN po.state != 'purchase' THEN 'no'
+                WHEN sub.lines_to_invoice > 0 THEN 'to invoice'
+                WHEN sub.lines = sub.lines_invoiced THEN 'invoiced'
+                ELSE 'no'
+            END
+            FROM (
+                SELECT
+                    po2.id,
+                    COUNT(pol1.*) AS lines,
+                    COUNT(pol2.*) AS lines_to_invoice,
+                    COUNT(pol3.*) AS lines_invoiced
+                FROM
+                    purchase_order po2
+                    -- pol1 contains all lines
+                    LEFT JOIN purchase_order_line pol1
+                        ON po2.id = pol1.order_id
+                    -- pol2 cotains lines to invoice
+                    LEFT JOIN purchse_order_line pol2
+                        ON po2.id = pol2.order_id AND
+                        ROUND(pol2.qty_invoiced, %(uom_precision)s) < ROUND(pol2.product_qty, %(uom_precision)s)
+                    -- pol3 contains invoiced lines
+                    LEFT JOIN purchse_order_line pol3
+                        ON po2.id = pol3.order_id AND
+                        ROUND(pol3.qty_invoiced, %(uom_precision)s) >= ROUND(pol3.product_qty, %(uom_precision)s)
+            ) AS sub
```

Comment:
Subqueries are usually slower. The approach in sale is not valid here?

Reply:
They are slower when in the `WHERE`, but in the `FROM` it shouldn't be a problem.